### PR TITLE
ember-cli

### LINKED ISSUE #3237: Move admin from EAK to Ember-Cli
The Ember App Kit has been deprecated in favor of [Ember-Cli](http://iamstef.net/ember-cli/). 

----------------------------------------------------------------------------

## Why
The EAK may be deprecated, but it's certainly working for us and will for the foreseeable future. So, why we should bother needs to be addressed

### Pros

1. Move client build steps out of grunt, woo for smaller gruntfile
2. faster, smarter builds via broccoli
3. hook into the growing ember-addon space (#3584 gets easier, for instance)
4. first class environment configuration
5. It's nice? I've been using it at work and liking it, apart from early life issues, but the core team is very responsive, with new releases nearly every other week

### Cons

1. Another build tool (broccoli)
2. contributors have to `npm install -g ember-cli bower grunt` now
3. i had to write a pros and cons list

## Blockers

These are the things that I think will make this hard, or that are on my mind

1. **Grunt vs Broccoli**

      `grunt dev` runs a watch command that recompiles server or client assets whenever a developer changes a file.  Moving the ember build step to an separate tool (broccoli) means that we need two processes/threads to do this, not just one (ie, `grunt dev` runs `ember build --watch`).  We could potentially leave the watching up to grunt and have it call `ember build` every time, but then we lose the benefit of broccoli intelligently rebuilding

2. ember-cli thinks it's the root of a project

    ember-cli expects jshint, package.json, and other project-level configurations to be placed at its directory, but that's obviously not going to be the case with Ghost. I'm assuming this can be configured

3. tests

    ember-cli tries its hardest to make sure you're doing tests, and comes with qunit. #2990 asks for a mocha based test framework, but I believe "anything's better than nothing" maybe the case here. There may also be some issues between our already running our own test. On the upside, it runs on phantomjs

4. base url

    ember-cli builds out a template with a `<base href="/your/settings" >` tag, but we obviously don't know where ghost will be served from. This should be a simple fix of removing the setup from ember-cli and doing it ourselves, save that ember-cli is using an express server's handlebars to build the tag, so adding our own handlebars to the template will possibly be odd?

5. eh, I think this list works for now. 

------------------------------------

I'll update as I start trying a switch

### PR DESCRIPTION
**Pull request by novaugust**
Closes #3237 
Moves Ghost's ember admin client from the deprecated EAK to the shiny new ember-cli
### Todos
- [x] get `ember serve` to show something, or at least print an error
- [x] move to ember-cli app layout
- [x] add client dependencies to ember-cli via bower / brocfile
- [x] move to ES6 for `ember` / `DS`
- [x] add sass/autoprfixer
- [x] asset minification
- [x] remove client related grunt tasks
- [x] update grunt to run ember-cli tasks
- [x] refactor client grunt tasks 
- [x] update development/release/install workflows and readmes (no change!!? pro users will probably want to `npm install -g ember-cli`, but people installing don't need to)
- [x] remove unused root-level deps in bower.json and package.json that have moved to the client-specific versions of those files
- [x] discuss new code styles now that we'll be able to do es6 (http://6to5.org/docs/learn-es6/)
- [x] get tests working (ping @jaswilli - would love to get some help here if you feel like playing)
#### Not being done in this PR
- Switch to ember-cli addons (simple-auth, etc)
- use es6 modules from deps that export them (brocfile tinkering)
### Todos from meeting 2-17
- [x] test release workflow (@erisds)
- [x] test different environments (dev vs prod)
- [x] remove client/docs
### Debuggery
- [x] fix error page assets

### CODE REVIEW

**Comment by novaugust** [comment]
Ghost's now trying to serve up the files generated by ember-cli.

I have an issue where my css lookups in `/ghost/*` are all getting redirected to the html `/ghost` that I need to figure out.

`301 REDIRECT /ghost/css/ghost.min.css?v=foobar -> /ghost/css/ghost.min.css/?v=foobar`
Not sure where that's from

Anyways, to play with this now, do
``sh
# in core/client
$ ember build --watch
```
```sh
# in the root dir, a separate console
$ node index
```

**Comment by novaugust** [comment]
Massively redid this, it actually works now ;) Updated todos

**Comment by novaugust** [comment]
Updated again, first pass through the gruntfile. `grunt [dev,init.prod]` are all working now, but `grunt dev` won't auto rebuild ember yet -- need to spawn a background `ember build --watch`

**Comment by jaswilli** [comment]
Awesome. I'm traveling right now but I'll try to give this a once-over in the next couple of days.

**Comment by novaugust** [comment]
Thanks!
And let's just agree not to merge any PRs that modify files in core/client
until this one's in, agreed?  :D

**Comment by jaswilli** [comment]
I've been going over this and it's definitely functional, which is really exciting.  I made some notes while going through it:

* do we need `client/tests/`?

* do we need the "qunit" dependencies in `client/bower.json`?

* status of `client/package.json`?
    * do we need things like express?

* remove client/app/config-dev.js and config-prod.js?

* do we need to consider ember-cli `disableAnalytics` (`client/.ember-cli`) in regards to `PRIVACY.md`?

* client lint files need to exclude `client/dist/`

* what are we doing about indentation (ember-cli generates indent of 2, everything else is 4)?

* bunch of client jscs/jshint errors that need to be dealt with

**Comment by novaugust** [comment]
> remove client/app/config-dev.js and config-prod.js
Yep, missed those this second time around.

> what are we doing about indentation (ember-cli generates indent of 2, everything else is 4)?
I'd vote for lazily bumping up to 4 - move to 4 when someone comes in and makes a change, but leave at 2 for now

> bunch of client jscs/jshint errors that need to be dealt with
Where can I see these?

> package.json
@rwjblue if we never use `ember serve`, do we need express in the ember-cli's package.json?

> disabling analytics
Yeah, i'll turn that off

> do we need `client/tests`
I was hoping you'd be able to tell me! I figured the unit tests you've been doing would go in there, but I'm sure we can keep doing them however we currently are (..?)

> q-unit deps
Probably not, if we aren't going to be using it ;)

**Comment by jaswilli** [comment]
> what are we doing about indentation (ember-cli generates indent of 2, everything else is 4)?
> I'd vote for lazily bumping up to 4 - move to 4 when someone comes in and makes a change, but leave at 2 for now

The only problem with that is that JSCS is checking for an indentation of 4.  So once the tests are running far enough to get to the linting, JSCS is going to go crazy over anything that was auto-generated with an indent of 2.

> bunch of client jscs/jshint errors that need to be dealt with
> Where can I see these?

If you fix up `Gruntfile.js` to exclude `client/dist` when building the list of files to lint, then you can run `grunt jshint:client` and `grunt jscs:client`.

> do we need client/tests
> I was hoping you'd be able to tell me! I figured the unit tests you've been doing would go in there, but I'm sure we can keep doing them however we currently are (..?)

I'll have to do some research on that.  I just wasn't sure if you had a strategy in place that involved leaving the ember-cli generated `client/tests/` in place but actually running our Ember tests from `core/test/client` (I realize they're not actually working right now).  I'm guessing the answer is "no" and `client/tests/` is just there because tests need to be sorted out still.

**Comment by novaugust** [comment]
>The only problem with that is that JSCS is checking for an indentation of 4. So once the tests are running far enough to get to the linting, JSCS is going to go crazy over anything that was auto-generated with an indent of 2.

Copy that, I'll just run the jshint and jscs tasks and clean up accordingly

> client/tests/

Yeah, ember-cli just put them there and I wasn't sure if we wanted to work with using it as our testing framework or keep doing our own thing. Looks like that's the last big decision that needs to be made.. mind if I lean on you for it? ^ ^

**Comment by rwjblue** [comment]
> @rwjblue if we never use ember serve, do we need express in the ember-cli's package.json?

Nope, you don't need it.

> Yeah, ember-cli just put them there and I wasn't sure if we wanted to work with using it as our testing framework or keep doing our own thing. Looks like that's the last big decision that needs to be made.. mind if I lean on you for it? ^ ^

In theory it would be better from ember-cli's perspective if the tests were in the "normal" location, but I think it is slightly worse ergonomics for y'all (meaning it would be nice if they could stay where they are currently).  We can make fixes/changes for this in ember-cli if needed, ping me in #ember-cli and we can chat through it...

> Do we need to consider ember-cli disableAnalytics (client/.ember-cli) in regards to PRIVACY.md?

Yes, that makes sense.

**Comment by jaswilli** [comment]
> In theory it would be better from ember-cli's perspective if the tests were in the "normal" location

I know we had this discussion a few months back regarding the location of Ember tests, and I was in favor of the path of least resistance then (i.e., locating them under `/client`), and I have to say that after working with ember-cli that I'm *really* in favor of doing it now.

I was able to get all the client tests running (and passing!) via `ember test` in like 20 minutes using ember-cli-mocha and ember-mocha.  I wasted that much time just trying to guess what to change in our current customized setup (and didn't get anything close to working).

**Comment by rwjblue** [comment]
> I know we had this discussion a few months back regarding the location of Ember tests, and I was in favor of the path of least resistance then (i.e., locating them under /client), and I have to say that after working with ember-cli that I'm really in favor of doing it now.

Less work for me, yay!

> Also, that weird problem I've been having where I can't test the rendered output of a component is gone when running the tests under ember-cli.

Awesome!

**Comment by jaswilli** [comment]
@novaugust 

I have the Ember tests working here: https://github.com/jaswilli/Ghost/tree/ember-cli-tests

I tried to create a discrete commit for every piece of work if you want to cherry-pick them onto your branch (or whatever works).

Basically what I did was:

* Set up the ember-cli test infrastructure to use Mocha.
* Removed all the qunit stuff.
* Moved the Ember tests from `core/test/client/unit/` into `core/client/tests/unit/`.
* Removed `core/test/client/`.
* Changed `Brocfile.js` to not run `jshint`.
* Re-worked the jshint / jscs grunt tasks to accommodate the ember-cli changes.
* Changed the indentation of ember-cli generated files from two to four spaces.
* Made the necessary adjustments to .jshintrc files and ES6 imports to the existing Ember test files.

**Comment by jaswilli** [comment]
All the tests are now passing.

**Comment by novaugust** [comment]
=O
So... squish commits (Leaving me first one + your first one with all the file moves alone) and merge after the next release?

**Comment by jaswilli** [comment]
Yeah, definitely want to leave the move/rename files commits standalone.  After that, I'd squash where it makes sense but I'd rather err on the side of more smaller commits than ending up with only a couple that are impossible to navigate.

**Comment by ErisDS** [comment]
Definitely don't stress about commit squashing on this one. When moving/renaming files you absolutely want to keep that separate from code changes. 

Where things go neatly together, squish em, otherwise don't.

**Comment by novaugust** [comment]
:+1: squashed one unneeded commit from mine work
... i can't believe this is happening

Thanks a lot for taking on the tests @jaswilli!

**Comment by novaugust** [comment]
rebased on 0.5.9. @ErisDS @jaswilli @dbalders shall we take it for one last spin and get it in? :)

**Comment by ErisDS** [comment]
I have taken this for a very quick spin, running `grunt init` + `npm start` or `npm start --production` seems to work, but running `grunt prod` makes both `npm start` and `npm start --production` throw this error:

> ERROR: Javascript files have not been built. 
 
Because the check at `core/server/index.js:82` fails due to all of the files having a hash in the filename:

http://puu.sh/gqnTa.png

Which lead me to wondering 2 things:

- what's that hash and why it's different for each file?
- where did the dev assets go?

In the existing system, `grunt init` builds all the standard (unconcatenated and unminified) assets for development mode, and `grunt prod` builds the production versions (but doesn't remove the development versions). If I run in production mode, the production assets should be used and if I run in development mode I get dev assets.

The intention of the separation is so that when running `grunt dev` (i.e. the watch task) you aren't waiting for slower minification functions to finish. 

I'm not sure what the ember-cli way to manage all of that is? Once I understand that, I can help get the release process working again. That relies heavily on the `.npmignore` file, which currently references the old project structure.

**Comment by novaugust** [comment]
The hashes come from the built in fingerprinting in ember-cli, easily turned off - sorry for not noticing them ;) http://www.ember-cli.com/#fingerprinting-and-cdn-urls
Re: minification, ember-cli already only minifies if the environment is production, so we're good there.

**Comment by novaugust** [comment]
Oh, and so you *do* know the ember-cli way of managing all that stuff, it's inside of the `Brocfile.js`, where we do `var app = new EmberApp(options) //or whatever it was`, we just mess with that options object

**Comment by novaugust** [comment]
Known issue:

If you `grunt init && grunt dev` for the very first time, the express server will fail because it will be checking for the built files before `ember build` has had time to output them...

File: Gruntfile.js [hunk 1]
```diff
@@ -31,54 +31,12 @@ var _              = require('lodash'),
                 return pattern.substr(1);
             }
             return '!' + pattern;
+        }).filter(function (pattern) {
+            // Remove empty patterns
+            return pattern !== '!'
```
Comment (rwjblue) [inline Gruntfile.js]: Missing semicolon...

Comment (novaugust) [inline Gruntfile.js]: Go home Robert. You're not wanted here.

**Comment by novaugust** [comment]
Found an issue with asset delivery for non-admin assets, leaving myself a note here:

screen shot 2015-03-11 at 18 22 20

**Comment by ErisDS** [comment]
@novaugust to fix that, I believe you just need to change `{{asset "css/ghost.min.css" ghost="true"}}` to `{{asset "ghost.css" ghost="true" minifyInProduction="true"}}` on L20 of user-error.hbs:

https://github.com/TryGhost/Ghost/blob/master/core/server/views/user-error.hbs#L20

**Comment by letsjustfixit** [comment]
Congratulations!

**Pull request merged**